keps/sig-apps/4017-pod-index-label/README.md

```diff
@@ -0,0 +1,898 @@
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+StatefulSets, similar to Indexed Jobs, assign a pod ordinal to each of their pods (this is
+analogous to the pod completion index for Indexed Jobs).
+
+Indexed Jobs set the pod index as an annotation, which means it can be utilized by the downward API,
+however, it cannot easily be used for selecting pods based on their index or filtering metrics/logs. 
+
+StatefulSets include the pod index as part of the pod name. This means that the only way for a 
+StatefulSet pod to know its own index is to [parse the pod name](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_utils.go#L63-L75),
+which is somewhat hacky solution. Furthermore, the pod index cannot be utilized by the downward API, 
+or for filtering metrics/logs.
+
+To address these issues, we propose to set the pod index as a **label** on pods for both StatefulSets and
+Indexed Jobs. This will allow the pod index to be utilized via the downward API, so StatefulSet pods
+can easily know their own index. In addition, this will allow metrics and logs to more easily be
+filtered by pod index, for both StatefulSets and Indexed Jobs.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Set pod index as label on pods for StatefulSet/Indexed Jobs.
+- Adding the label should not be disruptive to existing workloads.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+At a high level, the proposal is to modify the StatefulSet and Job controllers to set the pod index
+as a pod label at pod creation time (for jobs, this would only apply to jobs in
+Indexed completion mode). The details of this are outlined in the Design Details section below.
+
+StatefulSet pod label: `statefulset.kubernetes.io/pod-ordinal`
+Indexed Job pod label: `batch.kubernetes.io/job-completion-index` (same as existing annotation)
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+
+#### Story 1
```

Comment:
Is it possible to do any of these stories yet?  They aren't really mentioned in the motivation as reasons why you would need this feature to enable stories like this?

Reply:
User story 1 is mentioned in the motivation section. User story 2 isn't, but I can add it to the motivation section as well.